thesis/03-implementation.tex

```diff
@@ -83,23 +83,491 @@ \subsubsection{Backend}
+    \includegraphics[width=\textwidth]{assets/3-designer-creator-flow.png}
+\end{figure}
+
+Until \textit{Save the Guide} step, all data and actions are done on the client-side. When \textit{Save the Guide} action is invoked:
+\begin{itemize}
+	\item if a user is logged in, a request is sent to the Community with all configuration attached, or
+	\item a file download is initiated.
+\end{itemize}
+
+\begin{figure}[H]
+\caption{A Guide creation request. Almost all information is encoded in the \texttt{.fold} file}
+\begin{lstlisting}
+{
+	"private":false,
+	"guide_file":"data:text/json;base64,<base64 encoded guide>",
+	"thumbnail":"data:text/json;base64,<base64 thumbnail>"
+}
+\end{lstlisting}
+\end{figure}
+
+\begin{figure}[H]
+	\caption{A Guide creation response. The Instruction has not been processed yet, therefore, a \texttt{animation\_file} is not available.}
+	\label{guide-response}
+\begin{lstlisting}
+{
+	"id":6,
+	"owner":1,
+	"name":"penguin",
+	"published_at":"2020-12-01T18:49:18.131223Z",
+	"steps":9,
+	"guide_file":"https://api.<domain>/uploads/guides/file.fold",
+	"thumbnail_file":"https://api.<domain>/uploads/thumbnails/thumbnail.png",
+	"animation_file":null,
+	"status":"QUE",
+	"private":false,
+	"solved":false,
+	"liked":false,
+	"owner_username":"maciekmm1"
+}
+\end{lstlisting}
+\end{figure}
+
+
+\textbf{Guide Viewer}
+
+The purpose of the \textbf{Guide Viewer} is to show animated steps fluently. It consists of a model, a list of steps, and playback controls.
+
+\begin{figure}[H]
+  \caption{An overview of the Guide Viewer.}
+  \label{3-guide-viewer}
+  \centering
+    \includegraphics[width=\textwidth]{assets/3-guide-viewer.png}
+\end{figure}
+
+When a Guide is opened it fetches the \texttt{animation\_file} from the server and renders it.
+The response conforms to the same model as the aforementioned(\ref{guide-response}) guide creation.
+
+\medskip
+Playback controls allow rewinding, forwarding, stopping, and pausing to make it as approachable as possible for the Folder. Guide Viewer also provides functionality to load an arbitrary \tech{.fold} file by clicking the file icon located in the top-left corner of the screen.
+
+
+\subsubsection{Back-end}
+
+\begin{figure}[H]
+  \caption{The Community application consists of two main sub-applications.}
+  \label{3-backend-community-components}
+  \centering
+    \includegraphics[width=\textwidth]{assets/3-community-application.png}
+\end{figure}
+
+The community component is built using Django framework and split into a couple of sub-applications:
+
+\begin{description}
+	\item[accounts] - responsible for functionality related to user management. 
+	\item[guides] - responsible for controlling access and running all business logic related to Instruction and Animation files.
+\end{description}
+
+\textbf{Guides}
+
+Guides is the main sub-application. It is responsible for handling guide creation, deletion, marking a guide as a favorite, and delegation of Guide processing tasks.
```

Comment:
Hmm. Just the line above this, there is a short sentence describing what guides is responsible for, and it does not sound like the same things at the first glance.